fast Cython based skeletonize implementation

csachs opened the pull request:
Hello,

while working on a recent image analysis project, I found that the skeletonize function from scikit-image was taking … quite too much time in my opinion. Looking into the issue, I found that the current implementation, using a lot of numpy functions, apparently performs poorly.

As I've seen no PRs or issues concerning that situation, I've re-implemented the function as Cython code.
While it would possibly be more 'normal' to implement Zhan-Suen thinning in a correlation-free approach when low-level speed is available, I just converted the correlation based approach with the LUT as used in the Python code, as this most closely mimicks the current code.

Speed-up is notable for me, around 10-13x:

```python
In [1]: from skimage.data import horse
   ...: from skimage.morphology._skeletonize import _fast_skeletonize, _slow_skeletonize
   ...: image = horse()[:, :, 0] == 0
[ ... ]

In [2]: (_fast_skeletonize(image) == _slow_skeletonize(image)).all()
Out[2]: True

In [3]: %%timeit
   ...: _slow_skeletonize(image)
   ...: 
1 loops, best of 3: 202 ms per loop

In [4]: %%timeit
   ...: _fast_skeletonize(image)
   ...: 
100 loops, best of 3: 17.8 ms per loop
```

In the current state of the PR, I left the old Python version of the function inside, renaming it _slow_skeleton. Depending on the project's desires, it could obviously be removed if it becomes unnecessary.

Tests pass for me. (Or at least: only those which fail in master keep failing ;))

What do you think about the patch?

Best regards,
Christian

jni commented:
Hey @csachs! This is really cool, thanks for a great contribution!

A few notes:

- The docstring isn't rendering correctly (hence some of the Travis failures). I think that's because of the missing newlines between sections.
- We generally prefer a meatier wrapper around the Cython function, because Cython functions have limited introspection. In this case, I would put the function and docstring in `_skeletonize.py`, including the initial tests for `image.ndim` and binary values. Then, call `_skeleton_cy` on the checked image.
- Remove the old Python implementation. If tests pass, we don't like to keep old code as a record. That's what git is for!
- Some things need rewording for clarity, but I'll comment on them once the PR is in a more final form. e.g. `for odd_loop in (True, False):`
- We are in the process of normalising all our coordinate use to `row, col` rather than `y, x`. See [here](http://scikit-image.org/docs/dev/user_guide/numpy_images.html#coordinate-conventions). Please change the code accordingly. I would also rename e.g. `ymax` to `nrows`.

Finally, I'm wondering whether you have any experience in skeletonizing in 3D? That's an addition I've been hoping for for a long time but haven't managed to set aside the time to do. I'd love to get that in!

Thank you again!

csachs commented:
Hello,

thanks for the feedback!

* Whoups, I didn't try to build the docs. I thought I copy'n'pasted it over, don't know what went wrong there.
 I have some problems building the docs, so I hope it's fine now and I'll wait for Travis …
* The Cython function is now called from the wrapper.
* The old function body is removed.
* Added as well, called it `first_pass` now … feel free to add more comments.
* I've modified variable naming accordingly. I'd suggest adding the naming link http://scikit-image.org/docs/dev/user_guide/numpy_images.html#coordinate-conventions to http://scikit-image.org/docs/stable/contribute.html#stylistic-guidelines possibly with some hints to the Cython-wrapper issue as well. I took a look at the Contribution section, but unfortunately I didn't read all the docs before PR'ing here.

I have no experience with 3D skeletonization (and currently no 3D datasets of my own). Browsing some papers recently, I guess a more elaborate, kernel based approach might be applicable for higher dimensional data. (Personally, I'm afraid that such generalization might easily sacrifice specialized (yet common) 2D performance. Nonetheless, 3D functions would be a great addition)

Regards,
Christian

jni commented on skimage/morphology/_skeletonize_cy.pyx:
```diff
@@ -3,7 +3,109 @@
+cimport numpy as cnp
+
+def _fast_skeletonize(image):
+    """Optimized parts of the Zhang-Suen skeletonization.
```
blank lines between paragraphs.

Also, can you add a reference to the original paper? I didn't know that's what this algorithm was called. References are done as follows:

```
Zhang-Suen skeletonization [1]_.

References
----------
.. [1] K Zhang and F Suen. Skeletonizing binary images. (1950) Journal of Awesome 1:1-10.
```

jni commented on skimage/morphology/_skeletonize_cy.pyx:
```diff
@@ -3,7 +3,109 @@
+    Performs a hard-coded correlation to assign every neighborhood of 8 a
+    unique number, which in turn is used in conjunction with a look up
+    table to select the appropriate thinning criteria.
+
```
... but only one blank line, not two. =)

jni commented on skimage/morphology/_skeletonize_cy.pyx:
```diff
@@ -3,7 +3,109 @@
+                        # if the condition is met, the pixel is removed (unset)
+                        if (first_pass and neighbors == 1) or\
+                           ((not first_pass) and neighbors == 2) or\
+                           neighbors == 3:
```
I think PEP8 prefers parens for line continuation, and *extra* indentation to distinguish from subsequent blocks. So:

```python
if ((neighbors == 1 and first_pass) or
        (neighbors == 2 and not first_pass) or
        (neighbors == 3)):
    cleaned_skeleton[row, col] = 0
```

jni commented:
@csachs I've added some very nitpicky comments, but otherwise this is a fantastic PR! Very clean, thanks!

I'll submit a PR to the contributing guide forthwith, as that was a great suggestion. =)

csachs commented:
Okay, done.
I copied the Zhang-Suen reference verbatim from the old function (the Python wrapper had it already) … but a word of warning:
I did not personally verify that the LUT implements Zhang-Suen thinning. I just reimplemented the Python version as Cython …
As I stated earlier, Zhang-Suen thinning is normally implemented by directly evaluating the connectedness of a neighborhood. With this implementation, each pixel neighborhood is converted into a unique index for the LUT, whose value is then used to judge whether a pixel should be removed or not. … I did not verify that the manual way and the LUT yield identical results … but that's how it was implemented before in skimage …

Regards,
Christian

csachs commented:
Oups, was totally not aware of this. Pushed.

jni commented:
Brilliant! Just waiting on Travis now... =)

csachs commented:
Umh, a single Travis run is failing on something (completely) unrelated? How is this supposed to happen?

jni commented:
Ugh. Travis has been pretty unreliable recently, not to mention slow. I'm just going to go ahead and merge this, I just wanted the main versions (2.7 and 3.4) to run without issue. Thank you!
